### ISSUE TITLE
OMP: Warning #190 because of fork not waiting for parallel region to end

### ISSUE THREAD

**Issue by moskomule** [pytorch/pytorch#19996]
## 🐛 Bug

After updating to v1.1.0, I get `OMP: Warning #190: Forking a process while a parallel region is active is potentially unsafe.` from DataLoader when `num_workers>1`. This was not the case when I used v1.0.1 and v1.0.0. 

## To Reproduce

```
train_loader, test_loader = cifar10_loaders(batch_size=32, num_workers=num_workers)
for i in range(100):
    for _ in zip(test_loader, train_loader):
        pass
```

* len(test_loader.dataset) < len(train_loader.dataset)
* When `num_workers==1`, no warning.
* I also tested mnist and got the same warning.

## Environment

PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.14.0

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.2.148
...

Nvidia driver version: 396.44
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.1

Versions of relevant libraries:
[pip3] numpy==1.16.3
[pip3] torch==1.1.0
[pip3] torchvision==0.2.2
[conda] blas                      1.0                         mkl
[conda] cuda92                    1.0                           0    pytorch
[conda] mkl                       2019.3                      199
[conda] mkl_fft                   1.0.12           py37ha843d7b_0
[conda] mkl_random                1.0.2            py37hd81dba3_0
[conda] pytorch                   1.1.0           py3.7_cuda9.0.176_cudnn7.5.1_0    pytorch
[conda] torchvision               0.2.2                      py_3    pytorch

## Additional context

The code above was run inside a docker container.

macOS version does not cause this warning.

cc @ezyang @gchanan @zou3519 @SsnL

**Comment by ezyang** [comment]
Do you have a self-contained script which reproduces this problem?

**Comment by ezyang** [comment]
cc @ilia-cher

**Comment by fmassa** [comment]
This might be related to https://github.com/pytorch/pytorch/pull/19011 , as we set `num_threads` to 1 in the DataLoader https://github.com/pytorch/pytorch/blob/9005a2c0fc2d981c65b388ccd2ec4be3e1106a66/torch/utils/data/_utils/worker.py#L72

**Comment by ilia-cher** [comment]
Interesting, generally speaking OMP does not work well with `fork`, this is a known limitation of OMP.
It might be a harmless warning if we always call `set_num_threads(1)`, but let me dig into it first.

**Comment by tstandley** [comment]
I'm having this issue too.

**Comment by isalirezag** [comment]
I have the same issue.
it disapears when i do `num_workers = 0,` in dataloader though

**Comment by jhagege** [comment]
+1

**Comment by moskomule** [comment]
Could you post a reproductive code if possible?

**Comment by moskomule** [comment]
Again I encounter this problem. This problem seems to occur when other training processes are running.

**Comment by solitarysandman** [comment]
Anyone come up with a solution yet?

**Comment by ilia-cher** [comment]
> Again I encounter this problem. This problem seems to occur when you run the reproduction code above in parallel.

@moskomule, could you share how did you make it to show warning, how did you run the code you mentioned in parallel?

Summarizing known points:
- `fork` should not be called after OpenMP's parallel features has been used; in other words, if you started using OpenMP you shouldn't call `fork`, calling `fork` first and then using OpenMP is fine (https://stackoverflow.com/questions/49049388/understanding-openmp-shortcomings-regarding-fork)
- DataLoader uses `fork` to implement data parallelism
- DataLoader's worker also calls `set_num_threads(1)` from its worker (https://github.com/pytorch/pytorch/blob/master/torch/utils/data/_utils/worker.py#L121)

As I understand atm, if you've used parallelism (OpenMP) and after that called DataLoader in the same process then there might be a problem if parallelism is used later (e.g. via explicit `torch.set_num_threads(X)` call later)

**Comment by moskomule** [comment]
Hi @ilia-cher
I just run the snippet in https://github.com/pytorch/pytorch/issues/19996#issuecomment-490501684 in parallel by using two different terminals on a single machine.

**Comment by jeroneandrews** [comment]
I've also started getting this warning when using num_workers > 1 in both the train and test loader.

**Comment by KokeCacao** [comment]
I am having the same issue. Is there a way to bypass the warning? 
`import warnings
warnings.filterwarnings("ignore")`
Does not work for this specific warning.

**Comment by amir-abdi** [comment]
Same error here.

**Comment by moskomule** [comment]
@KokeCacao I don't think this warning is raised by Python, rather the C++ backend.

**Comment by ilia-cher** [comment]
it's hard to reproduce, I'm trying @moskomule example but still can't get a warning

**Comment by KokeCacao** [comment]
**For the reproduction ideas:** I discovered that this issue no longer exists after I changed my network to a smaller one. I suspect that it has to do with processing power?

**Comment by ilia-cher** [comment]
I think this is not related to the model itself, but to the OpenMP API calls we issue (e.g. set_num_threads), I'm still looking for the potential issues here and will comment with some findings.

**Comment by brynhayder** [comment]
I am also having this issue. It persists across data (MNIST, CIFAR10) and various architectures. I do not get the warning if I set `pin_memory=False`. I think it might be due to having three Dataloaders in my script like @FunkyKoki (i.e. train, test, validation) and iterating over two of them (test, val) while inside the loop of the other (train).

**Comment by jin8** [comment]
> Same error here. I create 3 dataset and load them by
> 
> `dataloader_1 = torch.utils.data.DataLoader(trainset_1, batch_size=arg.batch_size, shuffle=True,
> num_workers=arg.num_workers, pin_memory=True)
> 
> dataloader_2 = torch.utils.data.DataLoader(trainset_2, batch_size=arg.batch_size, shuffle=True,
> num_workers=arg.num_workers, pin_memory=True)
> 
> dataloader_3 = torch.utils.data.DataLoader(trainset_3, batch_size=arg.batch_size, shuffle=True,
> num_workers=arg.num_workers, pin_memory=True)`
> 
> And then the error comes out.

I had the same error after i updated my pytorch version to 1.2.0. So I downgraded to my previous version 1.0.1. I am no longer receiving the warning.

**Comment by peterbell10** [comment]
> I had the same error after i updated my pytorch version to 1.2.0. So I downgraded to my previous version 1.0.1. I am no longer receiving the warning.

Possibly related to https://github.com/pytorch/pytorch/issues/24080#issuecomment-535415022, in 1.0.1 torch copy wasn't properly multithreaded so OpenMP would not have been initialized when performing the copy in `pin_memory`.
https://github.com/pytorch/pytorch/blob/d3cab6571ed0d7efe9b5e01b2d8ca5993faab17a/aten/src/ATen/native/Memory.cpp#L31

**Comment by ngoldbaum** [comment]
> I just run the snippet in #19996 (comment) in parallel by using two different terminals on a single machine.

I just tried this and was unable to reproduce the OpenMP warning on my machine (Ubuntu 18.04 laptop with intel i7-8565U CPU, 4 physical cores).

**Comment by ezyang** [comment]
Maybe it's a 16.04 specific problem? Grasping at straws here though

**Comment by SsnL** [comment]
I can reproduce reliably on 18.04. The problem, I believe comes from when you have the `pin_memory_thread` doing parallelized `copy_` and a new data loader iterator starts, and `fork`s worker processes.

**Comment by SsnL** [comment]
If my assessment above is correct, then the warning should be mostly eliminated by https://github.com/pytorch/pytorch/pull/25111 . If it isn't, https://github.com/pytorch/pytorch/blob/515e3b85da93cd6ed3293bff3a392e6d8e271782/aten/src/ATen/ParallelOpenMP.h#L28 may need to be updated to check `get_num_threads() == 1`.

However, the fundamental issue that `fork` may happen when we are doing omp parallelization is not solved. In principle, we should add a function that blocks until all openmp active regions finish, and 
1. for python >= 3.7, call it in [`os.register_at_fork`](https://docs.python.org/3/library/os.html#os.register_at_fork) as the before fork hook.
2. for python < 3.7, at least manually call it before data loader workers if `fork` start method is used.

**Comment by SsnL** [comment]
Repro:
```py
import threading
import torch
import torchvision

x = torch.empty(300 * 300, 300 * 100)

e = threading.Event()

def f():
    while not e.is_set():
        x * 2

t = threading.Thread(target=f)
t.daemon = True

dl = torch.utils.data.DataLoader(torchvision.datasets.MNIST('./data', transform=torchvision.transforms.ToTensor(), download=True), batch_size=3, num_workers=4)

t.start()

next(iter(dl))

e.set()

t.join()
```

**Comment by SsnL** [comment]
Since the most common case to trigger the warning has been addressed, I'll remove high-pri flag now.

**Comment by Samleo8** [comment]
> Same error here. I create 3 dataset and load them by
> 
> ```python
> dataloader_1 = torch.utils.data.DataLoader(trainset_1, batch_size=arg.batch_size, shuffle=True,
>                                                    num_workers=arg.num_workers, pin_memory=True)
> 
> dataloader_2 = torch.utils.data.DataLoader(trainset_2, batch_size=arg.batch_size, shuffle=True,
>                                                    num_workers=arg.num_workers, pin_memory=True)
> 
> dataloader_3 = torch.utils.data.DataLoader(trainset_3, batch_size=arg.batch_size, shuffle=True,
>                                                    num_workers=arg.num_workers, pin_memory=True)
> ```
> 
> And then the error comes out.

Facing a similar issue when using multiple Dataloaders. What seems to happen is that once one of the processes from `dataloader_1` is finished, it starts a process for `dataloader_2`. For some reason, it causes my training to be stuck in a wait loop, and renders training/eval invalid. 

Is there a fix or workaround for this?
